internal/cluster/refresh_status_preprocessor.go

```diff
@@ -124,6 +141,103 @@ func (r *refreshPreprocessor) preprocess(ctx context.Context, c *clusterPreproce
 	return stateMachineInput, validationsOutput, nil
 }
 
+// recalculateOperatorDependencies calculates the operator dependencies and updates the database and the passed cluster
+// accordingly.
+func (r *refreshPreprocessor) recalculateOperatorDependencies(c *clusterPreprocessContext) error {
+	// Calculate and save the operators that have been added, updated or deleted:
+	previousOperators := c.cluster.MonitoredOperators
+	currentOperators, err := r.operatorsAPI.ResolveDependencies(c.cluster, c.cluster.MonitoredOperators)
+	if err != nil {
+		return errors.Wrapf(
+			err,
+			"failed to resolve operator dependencies for cluster '%s'",
+			c.clusterId,
+		)
+	}
+	var addedOperators, updatedOperators, deletedOperators []*models.MonitoredOperator
+	for _, currentOperator := range currentOperators {
+		if currentOperator.ClusterID == "" {
+			currentOperator.ClusterID = c.clusterId
+		}
+		previousOperator := operatorcommon.GetOperator(previousOperators, currentOperator.Name)
+		if previousOperator != nil {
+			if !reflect.DeepEqual(currentOperator, previousOperator) {
+				updatedOperators = append(updatedOperators, currentOperator)
+			}
+		} else {
+			addedOperators = append(addedOperators, currentOperator)
+		}
+	}
+	for _, previousOperator := range previousOperators {
+		if !operatorcommon.HasOperator(currentOperators, previousOperator.Name) {
+			deletedOperators = append(deletedOperators, previousOperator)
+		}
+	}
+	for _, addedOperator := range addedOperators {
+		err = c.db.Save(addedOperator).Error
```

Comment:
@rccrdpccl in the [previous](https://github.com/openshift/assisted-service/pull/7206#discussion_r1923262118) version of this pull request (now closed) @gamli75 was concerned 
that saving the cluster changes directly to DB will cause inconsistent values in the elastic. I added code below to also update the feature set. What else needs to be done to ensure consistency?

Reply:
In order to keep elastic data consistent, we notify about important (cluster, hosts and infraenv, only fields we deem important as we want to keep event volume contained) state changes. Unfortunately, there isn't a centralized place for this, so the code used to do this is dispersed.
An example could be [updating hosts](https://github.com/openshift/assisted-service/pull/7227#discussion_r1930284119) in some circumstances, but in this case we should [update the cluster state](https://github.com/openshift/assisted-service/blob/398cf47615a80714a11c2bd722681fc8c70a7cc4/internal/cluster/cluster.go#L294).
Related: https://github.com/openshift/assisted-service/blob/398cf47615a80714a11c2bd722681fc8c70a7cc4/internal/stream/notification_stream.go

Reply:
Added code to send the notifications, please take another look @rccrdpccl.